tests/test_sip_scenario.py

```diff
@@ -43,66 +42,21 @@
     await app.close()
 
 
-async def test_notify_with_router(test_server, protocol, loop, from_details, to_details):
-    notify_list = [0, 1, 2, 3, 4]
-    received_notify_futures = [loop.create_future() for _ in notify_list]
-    subscribe_future = loop.create_future()
-
-    async def subscribe(dialog, request):
-        assert len(dialog.peer.subscriber) == 1
-        await dialog.reply(request, status_code=200)
-        await asyncio.sleep(0.1)
-
-        for i in notify_list:
-            await dialog.notify(payload=str(i))
-        subscribe_future.set_result(None)
-
-    async def notify(dialog, request):
-        await dialog.reply(request, status_code=200)
-        received_notify_futures[int(request.payload)].set_result(request)
-
-    app = aiosip.Application(loop=loop)
-    server_app = aiosip.Application(loop=loop)
-    server_app.dialplan.add_user('pytest', {'SUBSCRIBE': subscribe})
-    server = await test_server(server_app)
-
-    peer = await app.connect(
-        protocol=protocol,
-        remote_addr=(server.sip_config['server_host'], server.sip_config['server_port'])
-    )
-
-    client_router = aiosip.Router()
-    client_router['notify'] = notify
-
-    await peer.subscribe(
-        expires=1800,
-        from_details=aiosip.Contact.from_header(from_details),
-        to_details=aiosip.Contact.from_header(to_details),
-        router=client_router
-    )
-
-    done, pending = await asyncio.wait(received_notify_futures, return_when=asyncio.ALL_COMPLETED, timeout=1)
-    received_notify = [f.result() for f in done]
-    assert len(pending) == 0
-
-    await subscribe_future
-    assert all((r.method == 'NOTIFY' for r in received_notify))
+async def test_authentication(test_server, protocol, loop, from_details, to_details):
+    password = 'abcdefg'
+    received_messages = list()
 
-    await server_app.close()
-    await app.close()
+    async def subscribe(request, message):
+        dialog = request._create_dialog()
 
+        received_messages.append(message)
+        assert not dialog.validate_auth(message, password)
+        await dialog.unauthorized(message)
 
-async def test_authentification(test_server, protocol, loop, from_details, to_details):
-    password = 'abcdefg'
-    received_request = list()
-
-    async def subscribe(dialog, request):
-        if dialog.validate_auth(request, password):
-            received_request.append(request)
-            await dialog.reply(request, 200)
-        else:
-            received_request.append(request)
-            await dialog.unauthorized(request)
+        async for message in dialog:
+            received_messages.append(message)
+            assert dialog.validate_auth(message, password)
+            await dialog.reply(message, 200)
```

Comment:
This is a practical example of where the previous server API was deficient. The auth challenge response should be coming in though the dialog, not the dialplan.